Initial implementation of Inference client

sivanantha321 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kserve/kserve/blob/master/docs/DEVELOPER_GUIDE.md
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Refer #3270

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Initial implementation of Inference client
```

sivanantha321 committed:
Initial implementation of inference client

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Add tests

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Use Inference client for e2e tests

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Upgrade pytest-asyncio to 0.23.4

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Fix mutable object initialization in default parameters

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Fix graph e2e tests

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Fix pmml test

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Add explain, support dict response, use inference client for internal requests

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Fix inference graph test and grpc headers

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Remove v1 datamodels

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Introduce protocol in client config

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

yuzisun commented on python/kserve/test/test_inference_client.py:
```diff
@@ -0,0 +1,167 @@
+from kserve.protocol.rest.server import RESTServer
+from test.test_server import DummyModel
+
+logging.basicConfig(level=logging.INFO)
```
Can we use kserve logger ?

sivanantha321 replied:
done

yuzisun commented on test/e2e/common/utils.py:
```diff
@@ -37,43 +39,63 @@
+grpc_client = None
 
-def grpc_stub(service_name, namespace):
+logging.basicConfig(level=logging.INFO)
```
Can we use kserve’s configured logger ?

sivanantha321 replied:
done

yuzisun commented on test/e2e/common/utils.py:
```diff
@@ -37,43 +39,63 @@
+        method_config = json.dumps({
+            "methodConfig": [{
+                "name": [{"service": "org.pytorch.serve.grpc.inference"}],
+                "retryPolicy": {
```
Can we move these configurations inside the client and set reasonable defaults ?

sivanantha321 replied:
Yeah, I am just using this for testing purposes.

sivanantha321 replied:
done

yuzisun commented on python/kserve/kserve/inference_client.py:
```diff
@@ -60,45 +93,365 @@ def __init__(self,
-              headers=None):
-        metadata = headers.items() if headers is not None else tuple()
+    async def infer(self,
+                    infer_request: Union[InferRequest, ModelInferRequest],
```
We do not need to expose the gRPC type ModelInferRequest as InferRequest can convert under the hood.

sivanantha321 replied:
done

yuzisun commented on python/kserve/kserve/inference_client.py:
```diff
@@ -60,45 +93,365 @@ def __init__(self,
+                    infer_request: Union[InferRequest, ModelInferRequest],
+                    client_timeout: Optional[float] = None,
+                    headers: Union[grpc.aio.Metadata, Sequence[Tuple[str, str]], None] = None) \
+            -> ModelInferResponse:
```
Same for the response.

sivanantha321 replied:
done

yuzisun commented on python/kserve/kserve/inference_client.py:
```diff
@@ -60,45 +93,365 @@ def __init__(self,
 
-    def infer(self,
-              infer_request: InferRequest,
-              client_timeout=None,
```
We probably want to set a reasonable default timeout like 60s here.

sivanantha321 replied:
done

sivanantha321 committed:
Support inference graph

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

remove logging configs

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Update default timeout to 60 seconds

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Add retry config for grpc client

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Fix infer model_name parameter

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Add tests for graph endpoint

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

debug

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

fix http client param mismatch

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

skip graph test

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

fix timeout in grpc client

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Fix url construction

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Fix explain

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
configure logger for e2e tests

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Rebase master

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Rebase master

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Fix grpc retry config

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Increase request timeout

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

yuzisun commented on python/kserve/kserve/inference_client.py:
```diff
@@ -64,40 +171,504 @@ def __init__(
+        self.verbose = verbose
+        if self.transport is None:
+            httpx.AsyncHTTPTransport(
+                retries=self.retries,
```
Can we retry based on the HTTP error code?

sivanantha321 replied:
httpx does not supports status code based retries. They recommend using other tools like tenacity. https://www.python-httpx.org/advanced/transports/#http-transport

yuzisun commented:
@sivanantha321 Can you help resolve the conflicts ? thanks

yuzisun commented on python/kserve/kserve/inference_client.py:
```diff
@@ -12,43 +12,150 @@
+                        break
+                if not is_exist:
+                    channel_opt.append(("grpc.enable_retries", 1))
+                    channel_opt.append(("grpc.service_config", service_config_json))
```
if `("grpc.enable_retries", 1)` exists in the channel args then it we append a duplicated entry ?

sivanantha321 replied:
Fixed.

yuzisun commented on python/kserve/kserve/inference_client.py:
```diff
@@ -64,40 +171,504 @@ def __init__(
+        if is_graph_endpoint:
+            return output
+        elif is_v2(self._config.protocol):
+            return InferResponse.from_rest(output.get("model_name"), response=output)
```
We might need an option to return raw dict for backwards compatibility.

yuzisun commented on python/kserve/kserve/model.py:
```diff
@@ -296,16 +274,7 @@ async def postprocess(
-                protocol, self.predictor_host, self.name
-            )
-
+    ) -> Union[Dict, InferResponse]:
```
Before it was only returning dict, this could be a breaking change if the transformer is depending on the dict type

sivanantha321 replied:
I don't think it will be a breaking change. This is a internal method used by predict function. Previously, Predict function gets the dict and converts it back to InferResponse, now the _http_predict directly returns the InferResponse

sivanantha321 commented:
/rerun-all

sivanantha321 commented:
/rerun-all

sivanantha321 committed:
configure logger for e2e tests

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Rebase master

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Rebase master

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Fix grpc retry config

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

Increase request timeout

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 commented:
/rerun-all

yuzisun commented on test/e2e/explainer/test_tabular_explainer.py:
Do we need this file ?

sivanantha321 replied:
Removed.

yuzisun commented on test/e2e/common/utils.py:
```diff
@@ -36,65 +40,60 @@
-        )
+def get_rest_client(protocol):
+    global rest_client_v1
+    global rest_client_v2
```
Let's avoid using global variables, you can create the client in a test fixture and pass to the test function

sivanantha321 replied:
Done

yuzisun commented:
@sivanantha321 Can you help resolve the conflict? a few small comments otherwise good to merge

sivanantha321 committed:
Rebase

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

sivanantha321 committed:
Use fixtures for rest client

Signed-off-by: Sivanantham Chinnaiyan <sivanantham.chinnaiyan@ideas2it.com>

yuzisun commented:
/lgtm
/approve
